diff --git a/arch/arm/mach-mx6/board-mx6q_var_som.c b/arch/arm/mach-mx6/board-mx6q_var_som.c
index 22f983e..8ac0a6b 100644
--- a/arch/arm/mach-mx6/board-mx6q_var_som.c
+++ b/arch/arm/mach-mx6/board-mx6q_var_som.c
@@ -87,6 +87,13 @@
 
 #define VAR_SOM_ECSPI3_CS0      IMX_GPIO_NR(4, 24)
 #define VAR_SOM_ADS7846_INT     IMX_GPIO_NR(4, 25)
+
+#define VAR_SOM_ECSPI1_CS0      IMX_GPIO_NR(4, 9)
+#define VAR_SOM_ECSPI1_CS1      IMX_GPIO_NR(4, 10)
+#define VAR_SOM_MAX1113X_INT    IMX_GPIO_NR(5, 12)
+#define VAR_SOM_MAX1113X_PD     IMX_GPIO_NR(5, 13)
+
+
 #define VAR_SOM_ADS7846_PD      IMX_GPIO_NR(4, 25)
 #define VAR_SOM_BACKLIGHT_EN    IMX_GPIO_NR(4, 30)
 
@@ -104,8 +111,11 @@
 #define VAR_SOM_PCIE_DIS_B      IMX_GPIO_NR(3, 21)
 #define VAR_SOM_PCIE_WAKE_B     IMX_GPIO_NR(3, 22)
 #define VAR_SOM_PCIE_PWR_EN     IMX_GPIO_NR(2, 12)
+#ifdef CONFIG_VAR_ORIGIN
 #define VAR_SOM_PCIE_RST_B_REVB IMX_GPIO_NR(5, 13)
-
+#else
+#define VAR_SOM_PCIE_RST_B_REVB IMX_GPIO_NR(4, 14)
+#endif
 #define VAR_SOM_KEY_BACK_GPIO   IMX_GPIO_NR(5, 20)
 
 #define VAR_SOM_WL1271_WL_IRQ   IMX_GPIO_NR(6, 17)
@@ -373,15 +383,35 @@ static struct fec_platform_data fec_data __initdata = {
 	.phy = PHY_INTERFACE_MODE_RGMII,
 };
 
-static int mx6q_var_som_spi_cs[] = {
+static int mx6q_var_som_spi0_cs[] = {
+	VAR_SOM_ECSPI1_CS0,
+	VAR_SOM_ECSPI1_CS1,
+};
+
+static int mx6q_var_som_spi1_cs[] = {
+};
+
+static int mx6q_var_som_spi2_cs[] = {
 	VAR_SOM_ECSPI3_CS0,
 };
 
-static const struct spi_imx_master mx6q_var_som_spi_data __initconst = {
-	.chipselect     = mx6q_var_som_spi_cs,
-	.num_chipselect = ARRAY_SIZE(mx6q_var_som_spi_cs),
+
+static const struct spi_imx_master mx6q_var_som_spi0_data __initconst = {
+	.chipselect     = mx6q_var_som_spi0_cs,
+	.num_chipselect = ARRAY_SIZE(mx6q_var_som_spi0_cs),
+};
+
+static const struct spi_imx_master mx6q_var_som_spi1_data __initconst = {
+	.chipselect     = mx6q_var_som_spi1_cs,
+	.num_chipselect = ARRAY_SIZE(mx6q_var_som_spi1_cs),
+};
+
+static const struct spi_imx_master mx6q_var_som_spi2_data __initconst = {
+	.chipselect     = mx6q_var_som_spi2_cs,
+	.num_chipselect = ARRAY_SIZE(mx6q_var_som_spi2_cs),
 };
 
+
 static struct ads7846_platform_data ads7846_config = {
 	.x_max               = 8080,
 	.y_max               = 7980,
@@ -409,12 +439,24 @@ static struct spi_board_info mx6_var_som_spi_ts_device[] __initdata = {
 	},
 };
 
+static struct spi_board_info mx6_var_som_spi_device[] __initdata = {
+	{
+		.modalias	= "max1113x",
+		.max_speed_hz	= 8000000, //48 Mbps
+		.bus_num	= 0,
+		.chip_select	= 1,
+		//.irq = gpio_to_irq(VAR_SOM_MAX1113X_INT),
+	},
+};
+
 static void spi_device_init(void)
 {
 	if (var_lcd_index != VAR_LCD_CTW6120) {
 		spi_register_board_info(mx6_var_som_spi_ts_device,
 				ARRAY_SIZE(mx6_var_som_spi_ts_device));
 	}
+	spi_register_board_info(mx6_var_som_spi_device,
+					ARRAY_SIZE(mx6_var_som_spi_device));
 }
 
 /* Audio
@@ -1163,7 +1205,9 @@ static void __init mx6_var_som_board_init(void)
 	gpio_direction_output(VAR_SOM_BACKLIGHT_EN, 1);
 
 	/* SPI */
-	imx6q_add_ecspi(2, &mx6q_var_som_spi_data);
+	imx6q_add_ecspi(0, &mx6q_var_som_spi0_data);
+	imx6q_add_ecspi(1, &mx6q_var_som_spi1_data);
+	imx6q_add_ecspi(2, &mx6q_var_som_spi2_data);
 	spi_device_init();
 
 	imx6q_add_mxc_hdmi(&hdmi_data);
diff --git a/arch/arm/mach-mx6/board-mx6solo_var_som.h b/arch/arm/mach-mx6/board-mx6solo_var_som.h
index 4f3cccb..2a91183 100644
--- a/arch/arm/mach-mx6/board-mx6solo_var_som.h
+++ b/arch/arm/mach-mx6/board-mx6solo_var_som.h
@@ -56,6 +56,17 @@ static iomux_v3_cfg_t mx6dl_var_som_pads[] = {
 	MX6DL_PAD_CSI0_DAT6__AUDMUX_AUD3_TXFS,
 	MX6DL_PAD_CSI0_DAT7__AUDMUX_AUD3_RXD,
 
+	/* SPI1 (adc max11136) */
+	MX6DL_PAD_DISP0_DAT22__ECSPI1_MISO,
+	MX6DL_PAD_DISP0_DAT21__ECSPI1_MOSI,
+	MX6DL_PAD_DISP0_DAT20__ECSPI1_SCLK,
+
+	MX6DL_PAD_KEY_COL2__ECSPI1_SS1,
+
+	//MX6DL_PAD_DISP0_DAT19__GPIO_5_13,
+	//MX6DL_PAD_DISP0_DAT18__GPIO_5_12,
+
+
 	/* SPI3 (ads7846) */
 	MX6DL_PAD_DISP0_DAT2__ECSPI3_MISO,
 	MX6DL_PAD_DISP0_DAT1__ECSPI3_MOSI,
@@ -139,7 +150,11 @@ static iomux_v3_cfg_t mx6dl_var_som_pads[] = {
 
 	/* PCIE */
 	MX6DL_PAD_EIM_D21__GPIO_3_21, /* PCIE_DIS */
+#ifdef CONFIG_VAR_ORIGIN
 	MX6DL_PAD_DISP0_DAT19__GPIO_5_13, /* PCIE_RST */
+#else
+	MX6DL_PAD_KEY_COL4__GPIO_4_14,	/* PCIE_RST */
+#endif
 	MX6DL_PAD_SD4_DAT4__GPIO_2_12, /* PCIE_PWR_ENA */ 
 
 	/* WLAN - USDHC3 */
