diff --git a/arch/arm/mach-mx6/board-mx6q_var_som.c b/arch/arm/mach-mx6/board-mx6q_var_som.c
index c869038..ab7f6e3 100644
--- a/arch/arm/mach-mx6/board-mx6q_var_som.c
+++ b/arch/arm/mach-mx6/board-mx6q_var_som.c
@@ -275,6 +275,24 @@ static struct esdhc_platform_data mx6q_var_som_sd3_data = {
 /* NAND */
 static struct mtd_partition mx6q_var_som_partitions[] = {
 	{
+		.name	= "spl",
+		.offset	= 0,
+		.size	= 0x00200000,
+	}, {
+		.name	= "bootloader",
+		.offset	= 0x00200000,
+		.size	= 0x00200000,
+	}, {
+		.name	= "kernel",
+		.offset	= 0x00400000,
+		.size	= 0x00600000,
+	}, {
+		.name	= "rootfs",
+		.offset	= 0x00a00000,
+		.size	= 0x3f600000,
+
+#if 0
+	{
 		.name	= "bootloader",
 		.offset	= 0,
 		.size	= 0x00200000,
@@ -286,6 +304,7 @@ static struct mtd_partition mx6q_var_som_partitions[] = {
 		.name	= "rootfs",
 		.offset	= MTDPART_OFS_APPEND,
 		.size	= MTDPART_SIZ_FULL,
+#endif
 #ifdef ANDROID_NAND_RECOVERY
 	}, {
 		.name	= "android_boot",
